### DISCUSSION TITLE
Using multiple TRPC clients in the same app

### DISCUSSION THREAD

**Discussion by dodas** [trpc/trpc#1990] [Q&A]
**👉👉 👉  Moved to #2121**

---

I am trying to have multiple `TRPCRouter`s on backend and separate `TRPCClient`s to access them.
Looking at [the way react hooks / contexts are set up](https://github.com/trpc/trpc/blob/main/packages/react/src/createReactQueryHooks.tsx), seems like there is no easy way to have 2 separate instances, since they reuse the same `TRPCContext` singleton.
I am expecting to be able to use these two routers/clients completely separately, without them conflicting with each other.
Is this a viable use-case? It seems that the current design does not allow this.

Thanks!

**Comment by KATT** [comment]
I'm on my phone so can't type much but see #1226

You can actually use the same client on the front-end even if you're talking to different backends

**Reply by KATT** [reply]
Glue them together with a virtual router and use a client-side link to route requests to the right one

**Comment by dodas** [comment]
I would really like to avoid that and actually keep them separated.
I am wrapping TRPC client into my own little library, basically to create an SDK, and would like to abstract away these details.

Would `react-query`'s API be a limitation as well if we were trying to allow for independent clients in trpc?

**Reply by KATT** [reply]
> I would really like to avoid that and actually keep them separated.

They are namespaced though, this is the "virtual server" that combines two servers that are deployed independently:

https://github.com/trpc/trpc/blob/e37b1ca34b2e5ecce1b016cd82dc19e882e3710e/examples/soa/src/server-_combined.ts#L6-L9

The client will know which server to call depending on the prefix:

https://github.com/trpc/trpc/blob/e37b1ca34b2e5ecce1b016cd82dc19e882e3710e/examples/soa/src/client.ts#L21-L47



> Would `react-query`'s API be a limitation as well if we were trying to allow for independent clients in trpc?

Unsure, never researched this.

**Reply by dodas** [reply]
I can't really do that.
I am wrapping TRPCClient into my own library (SDK), which is hiding its implementation details.
Now imagine someone uses my SDK along with a trpc itself (or other sdk that uses trpc) – they all need to work independently, without the consumer needing to create a "virtual server" on top of them.

**Reply by KATT** [reply]
Hmm, I see, this is more of a react-query question then really, I am unsure how to answer this.

You could do your own context provider and only use the raw client too, but you'll miss some niceities from rq

**Reply by dodas** [reply]
Just found out `react-query@4.0` supports exactly this use-case, see [Custom Contexts for Multiple Providers](https://github.com/TanStack/query/blob/beta/docs/guides/migrating-to-react-query-4.md#custom-contexts-for-multiple-providers).
So `react-query` now supports this, user would just need to set up his `QueryClientProvider` in such way.
However, I think we will still need to update [this](https://github.com/trpc/trpc/blob/899b264ab704e5eed22086fe35fb68dc751e9c5e/packages/react/src/createReactQueryHooks.tsx#L113) in `@trpc/react` so we create a new trpc context for each `createReactQueryHooks()` call, instead of reusing the singleton.

**Reply by dodas** [reply]
Just tested it:
I patched `@trpc/react` to move the `createContext(null)` call inside the `createReactQueryHooks()` and set up my `QueryClientProvider` to pass in custom context – everything is working as expected.

Are you willing to accept a PR with this change?

**Comment by dodas** [comment]
@KATT bump :) thoughts?

**Reply by KATT** [reply]
I'm not sure what's needed on tRPC's end..?

**Reply by dodas** [reply]
I described it [above](https://github.com/trpc/trpc/discussions/1990#discussioncomment-2949992) – we need to not reuse the same TRPCContext for all clients, but rather create a new context for each `createReactQueryHooks` call. That way the hooks returned will use the corresponding `TRPCClient`/`TRPCContext`, rather than any TRPC provider that is found up the tree.

**Reply by KATT** [reply]
This is an edge-case that I don't have bandwidth for to work on any time soon but feel to make a PR towards the `next`-branch.

**Reply by KATT** [reply]
Happy to give guidance too, feel free to @ me on Discord if you want to contribute.

**Reply by mmkal** [reply]
I’d also be interested in this feature (but wouldn’t be able to write it any time soon) - so happy to be another person to bounce ideas off if you want to work on it @dodas
